usr/src/lib/libzfs/common/libzfs_sendrecv.c

```diff
@@ -2797,7 +2797,7 @@ zfs_receive_package(libzfs_handle_t *hdl, int fd, const char *destname,
 				*strchr(tofs, '@') = '\0';
 			}
 
-			if (recursive && !flags->dryrun && !flags->nomount) {
+			if (!flags->dryrun && !flags->nomount) {
```

Comment:
I agree the second condition is redundant and could be misleading, that's why I'm saying we should remove it.  To summarize:

```
if (... && recursive) {
...
  if (recursive & ...) {
...
  }
}
```
should be:
```
if (... && recursive) {
...
  if (...) {
...
  }
}
```